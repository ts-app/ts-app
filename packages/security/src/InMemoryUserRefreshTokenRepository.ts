import { of } from 'rxjs'
import { UserRefreshTokenRepository } from './UserRefreshTokenRepository'

export class InMemoryUserRefreshTokenRepository extends UserRefreshTokenRepository {
  tokens: {
    [ key: string ]: Set<string>
  } = {}

  register (userId: string, refreshToken: string) {
    if (!this.tokens[ userId ]) {
      this.tokens[ userId ] = new Set()
    }

    this.tokens[ userId ].add(refreshToken)
    return of(null)
  }

  getTokens (userId: string) {
    if (!this.tokens[ userId ]) {
      this.tokens[ userId ] = new Set()
    }

    return of(this.tokens[ userId ])
  }

  clearTokens (userId?: string) {
    if (!userId) {
      this.tokens = {}
    } else {
      const userTokens = this.tokens[ userId ]
      if (userTokens) {
        userTokens.clear()
      }
    }
    return of(null)
  }
}
