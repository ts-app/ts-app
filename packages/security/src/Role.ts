import { Timestampable } from '@ts-app/common'

export interface Role extends Timestampable {
  id: string
  name: string
}
