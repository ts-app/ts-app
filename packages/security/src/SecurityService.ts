import { Observable } from 'rxjs'
import { User, UserProfile, BusinessError, FindInput, FindOutput } from '@ts-app/common'

export abstract class SecurityService {
  abstract signUp (input: { email: string, password: string }): Observable<{
    user?: User
    error?: BusinessError
  }>

  abstract loginWithEmailPassword (input: { email: string, password: string }): Observable<{
    user?: User
    accessToken?: string
    refreshToken?: string
    error?: BusinessError
  }>

  abstract loginWithRefreshToken (refreshToken: string): Observable<{
    user?: User
    accessToken?: string
    refreshToken?: string
  }>

  abstract user (id: string): Observable<User | null>

  abstract users (input: FindInput): Observable<FindOutput<User>>

  abstract userByEmail (email: string): Observable<User | null>

  abstract removeUser (id: string): Observable<null>

  abstract seedUsers (input: {
    force?: boolean
    userCount?: number
  }): Observable<null>

  abstract updateProfile<T extends UserProfile> (input: {
    id: string
    profile: Partial<T>
  }): Observable<null>

  abstract reset (): Observable<null>
}
