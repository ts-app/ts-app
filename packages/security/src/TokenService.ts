import * as crypto from 'crypto'
import * as jwt from 'jsonwebtoken'

type VerifyPayload = {
  payload: any
  userId: string
  exp: number
  iat: number
  jti: number
}
type GenerateOptions = {
  secretOrPrivateKey?: string
  accessTokenExpiresIn?: string
  refreshTokenExpiresIn?: string
}
type VerifyOptions = {
  secretOrPublicKey?: string
}

export class TokenService {
  constructor (
    private secretOrPrivateKey: string,
    private secretOrPublicKey: string,
    private accessTokenExpiresIn = '10m',
    private refreshTokenExpiresIn = '30d') {

  }

  generateUserToken (userId: string, payload: { [ key: string ]: any }, options: GenerateOptions = {}) {
    const secretOrKey = options.secretOrPrivateKey || this.secretOrPrivateKey
    const accessTokenExpiresIn = options.accessTokenExpiresIn || this.accessTokenExpiresIn
    const refreshTokenExpiresIn = options.refreshTokenExpiresIn || this.refreshTokenExpiresIn

    const accessToken = jwt.sign({ payload, userId }, secretOrKey, {
      expiresIn: accessTokenExpiresIn,
      jwtid: this.uuid()
    })
    const refreshToken = jwt.sign({ payload, userId }, secretOrKey, {
      expiresIn: refreshTokenExpiresIn,
      jwtid: this.uuid()
    })

    return { accessToken, refreshToken }
  }

  verify (token: string, options: VerifyOptions = {}) {
    return jwt.verify(token, options.secretOrPublicKey || this.secretOrPublicKey) as VerifyPayload
  }

  private uuid () {
    return crypto.randomBytes(16).toString('hex')
  }
}
