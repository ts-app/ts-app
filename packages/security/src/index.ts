export * from './MongoSecurityService'
export * from './SecurityService'
