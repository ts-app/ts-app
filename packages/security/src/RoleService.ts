import { FindInput, FindOutput, User } from '@ts-app/common'
import { Role } from './Role'
import { Observable } from 'rxjs'

/**
 * Role service allows assignment of roles to users for authorization purposes. It also provides
 * functions to check roles assigned to user, based on groups.
 *
 * Credits: This interface was heavily inspired by alanning/roles-npm.
 */
export abstract class RoleService {
  static GLOBAL = '_global_'

  /**
   * Add users to specified roles, optionally specifying a group. If group name is not specified,
   * roles are assigned to RoleServiceConstants.GLOBAL group.
   */
  abstract addUsersToRoles (input: {
    userIds: string[], roles: string[],
    group?: string
  }): Observable<null>

  /**
   * Create specified role name.
   */
  abstract createRole (name: string): Observable<string>

  /**
   * Delete specified role name.
   *
   * Returns true if the specified role name exist, is not used and was removed. Otherwise, this
   * function returns false.
   */
  abstract removeRole (name: string): Observable<boolean>

  /**
   * Get user's groups where user has role(s) assigned. RoleServiceConstants.GLOBAL will be omitted.
   *
   * If role is specified, restrict groups returned to those with specified role assigned.
   */
  abstract getGroupsForUser (input: { userId: string, role?: string }): Observable<string[]>

  /**
   * Find roles with option to filter, limit and perform pagination on results.
   */
  abstract findRoles (input: FindInput): Observable<FindOutput<Role>>

  /**
   * Get roles assigned to user, optionally restricting specifying a group. Roles assigned to
   * RoleServiceConstants.GLOBAL will be included.
   *
   * If group name is not specified, all roles including those assigned to
   * RoleServiceConstants.GLOBAL are returned.
   */
  abstract getRolesForUser (input: { userId: string, group?: string }): Observable<string[]>

  /**
   * Get users with specified role(s) and group. User is considered a match if at least one role
   * matches.
   */
  abstract getUsersInRoles (input: { roles: string[], group?: string, limit?: number, cursor?: string }): Observable<FindOutput<User>>

  /**
   * Unassign user(s) from specified role(s).
   */
  abstract removeUsersFromRoles (input: { userIds: string[], roles: string[], group?: string }): Observable<null>

  /**
   * Remove all assigned roles from specified user(s).
   */
  abstract removeUsersFromAllRoles (input: { userIds: string[] }): Observable<null>

  /**
   * Returns true if user is assigned to one or more of the specified rule(s). This can be further
   * restricted based on group name.
   */
  abstract isUserInRoles (input: { userId: string, roles: string[], group?: string }): Observable<boolean>

  /**
   * Get role with the specified ID.
   *
   * @param id
   */
  abstract role (id: string): Observable<Role | null>

  abstract updateRole (id: string, role: Partial<Role>): Observable<null>
}
