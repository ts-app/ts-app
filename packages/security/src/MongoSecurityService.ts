import * as bcryptjs from 'bcryptjs'
import { SecurityService } from './SecurityService'
import { of, Observable, from, throwError, range } from 'rxjs'
import {
  makeTimestampable,
  ofBusinessError,
  escapeRegex,
  validateEmail,
  User,
  BusinessError,
  FindInput,
  FindOutput,
  UserProfile
} from '@ts-app/common'
import { MongoService } from '@ts-app/mongo'
import { concatMap, map, mapTo, toArray } from 'rxjs/operators'

export class MongoSecurityService extends SecurityService {
  constructor (private mongoService: MongoService) {
    super()
  }

  signUp (input: { email: string; password: string }): Observable<{ user?: User; error?: BusinessError }> {
    if (!validateEmail(input.email)) {
      return ofBusinessError('Invalid email')
    }

    return this.mongoService.get('users', { 'emails.email': input.email }).pipe(
      concatMap(emailExist => {
        if (emailExist) {
          return ofBusinessError('User already exist')
        } else {
          return from(bcryptjs.hash(input.password, 8)).pipe(
            concatMap(bcrypt => {
              const makeUser = (bcrypt: string) => ({
                ...makeTimestampable(),
                services: {
                  password: { bcrypt }
                },
                emails: [
                  {
                    email: input.email,
                    verified: true
                  }
                ],
                profile: {
                  email: input.email,
                  displayName: input.email.substr(0, input.email.indexOf('@'))
                }
              })
              return this.mongoService.create('users', makeUser(bcrypt))
            }),
            concatMap(id => this.mongoService.get('users', id)),
            map(user => user ? { user: user } : { error: `Error loading user [${input.email}]` })
          )
        }
      })
    )
  }

  loginWithEmailPassword (input: { email: string, password: string }):
    Observable<{ user?: User, accessToken?: string, refreshToken?: string, error?: BusinessError }> {
    const filter = {
      emails: {
        email: input.email,
        verified: true
      }
    }

    return this.mongoService.get<User>('users', filter).pipe(
      concatMap(user => {
        const userBcrypt = user && user.services && user.services.password && user.services.password.bcrypt
        if (userBcrypt) {
          return from(bcryptjs.compare(input.password, userBcrypt)).pipe(
            map(passwordMatch => ({ passwordMatch, user }))
          )
        } else {
          return throwError('User password is not set.')
        }
      }),
      concatMap<{ passwordMatch: boolean, user: User }, { user: User }>(({ passwordMatch, user }) => {
        if (passwordMatch) {
          return of({ user })
        } else {
          return ofBusinessError('Invalid login attempt')
        }
      })
    )
  }

  user (id: string): Observable<User | null> {
    return this.mongoService.get<User>('users', id)
  }

  users (input: FindInput): Observable<FindOutput<User>> {
    let filter = {}
    let { sort } = input
    const { q, limit, cursor, project } = input
    if (q && q.trim().length > 0) {
      filter = {
        $or: [
          { 'profile.email': { $regex: `^${escapeRegex(q)}`, $options: 'i' } },
          { 'profile.email': { $regex: `${escapeRegex(q)}$`, $options: 'i' } },

          { 'profile.displayName': { $regex: `^${escapeRegex(q)}`, $options: 'i' } },
          { 'profile.displayName': { $regex: `${escapeRegex(q)}$`, $options: 'i' } }
        ]
      }
    }

    if (!sort) {
      sort = [ { field: 'profile.displayName', asc: true } ]
    }

    return this.mongoService.findWithCursor('users', filter, limit, cursor, sort, project)
  }

  removeUser (id: string): Observable<null> {
    return this.mongoService.remove('users', id).pipe(mapTo(null))
  }

  seedUsers (input: { force?: boolean; userCount?: number }): Observable<null> {
    return of(input.force).pipe(
      // --- if forced, remove any existing seeded test users
      concatMap(force => {
        if (force) {
          return this.mongoService.remove('users', {
            $or: [
              { 'emails.email': { $regex: 'user[0-9]+@test\.local' } },
              { 'emails.email': 'admin@test.local' }
            ]
          })
        }
        return of(null)
      }),
      // --- prevent seeding if users exist (force just remove previously seeded users)
      concatMap(() => this.mongoService.count('users')),
      concatMap(userCount => userCount > 0 ? throwError('Cannot seed database with users') : of(null)),
      // --- seed users
      // create admin
      concatMap(() => this.signUp({ email: 'admin@test.local', password: 'testAdmin' })),
      // throw error if admin sign up error
      concatMap(signUp => signUp.error ? throwError(signUp.error) : of(null)),
      // seed 10 users
      concatMap(() => range(1, 10)),
      concatMap(no => this.signUp({
        email: `user${no}@test.local`,
        password: `testUser${no}`
      })),
      // throw error if user sign up error
      concatMap(signUp => {
        return signUp.error ? throwError(signUp.error) : of(null)
      }),
      toArray(),
      mapTo(null)
    )
  }

  updateProfile<T extends UserProfile> (input: { id: string; profile: Partial<T> }): Observable<null> {
    const profileWithPrefix = Object.keys(input.profile).reduce((p, key) => {
      p[ `profile.${key}` ] = (input.profile as any)[ key ]
      return p
    }, {} as any)

    return this.mongoService.update('users', input.id, {
      $set: {
        ...profileWithPrefix
      }
    })
  }

  reset (): Observable<null> {
    return this.mongoService.findWithCursor<User>('users', {
      'emails.email': 'admin@test.local'
    }).pipe(
      concatMap(users => {
        if (users.docs.length === 0) {
          return throwError('Pre-requisite to reset not met')
        }

        return this.mongoService.dropCollection('users').pipe(mapTo(null))
      })
    )
  }
}
