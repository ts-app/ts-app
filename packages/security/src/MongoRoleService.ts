import { MongoService } from '@ts-app/mongo'
import { assert, FindInput, FindOutput, User, makeTimestampable } from '@ts-app/common'
import { Observable, of, throwError } from 'rxjs'
import { concatMap, mapTo, map, catchError, toArray } from 'rxjs/operators'
import { RoleService } from './RoleService'
import { Role } from './Role'
import { ObjectId } from 'bson'

export class MongoRoleService extends RoleService {
  constructor (private mongoService: MongoService) {
    super()
  }

  addUsersToRoles (input: { userIds: string[]; roles: string[]; group?: string }): Observable<null> {
    input.group = input.group || MongoRoleService.GLOBAL.replace(/\./g, '_')

    if (input.group.startsWith('$')) {
      return throwError('Group cannot start with \'$\'')
    }

    return this.mongoService.cursor<Role>('roles', { name: { $in: input.roles } }).pipe(
      concatMap(cursor => cursor.toArray()),
      // --- create missing role(s) in role collection
      concatMap(matchingRolesInDb => {
        const rolesNotInDb = input.roles.filter(roleName => !matchingRolesInDb.find(roleInDb => roleInDb.name === roleName))
        if (rolesNotInDb) {
          return of(...rolesNotInDb).pipe(
            concatMap(roleName => this.createRole(roleName))
          ).pipe(toArray())
        } else {
          return of(null)
        }
      }),

      // --- assign roles to users collection
      concatMap(() => this.mongoService.collection('users')),

      concatMap(Users => Users.updateMany(
        {
          _id: { $in: input.userIds.map(id => new ObjectId(id)) }
        },
        {
          $addToSet: {
            roles: {
              $each: input.roles.map(role => ({ role, group: input.group }))
            }
          }
        })),
      mapTo(null)
    )
  }

  createRole (name: string): Observable<string> {
    assert(!!name, 'Role name required')

    return this.mongoService.create<Role>(
      'roles',
      { name, ...makeTimestampable() }
    )
  }

  findRoles (input: FindInput): Observable<FindOutput<Role>> {
    return this.mongoService.find('roles', input, [ 'name' ])
  }

  getGroupsForUser (input: { userId: string; role?: string }): Observable<string[]> {
    let filter
    if (input.role) {
      filter = {
        _id: new ObjectId(input.userId),
        'roles.role': input.role
      }
    } else {
      filter = { _id: new ObjectId(input.userId) }
    }

    return this.mongoService.get<User>('users', filter, {
      fields: { roles: 1 }
    }).pipe(
      concatMap(user => {
        if (!user) {
          // error if cannot find user, return empty array
          return of([])
        } else if (input.role) {
          // return groups where user has specified role assigned (except global)
          const groupNames = user.roles
            .filter(currentRole => currentRole.group !== RoleService.GLOBAL && currentRole.role === input.role)
            .map(currentRole => currentRole.group)

          return of(Array.from(new Set(groupNames)))
        } else {
          // role not specified, return all groups assigned to user (except global)
          const groupNames = user.roles.filter(currentRole => currentRole.group !== RoleService.GLOBAL)
            .map(currentRole => currentRole.group)

          return of(Array.from(new Set(groupNames)))
        }
      })
    )
  }

  getRolesForUser (input: { userId: string; group?: string }): Observable<string[]> {
    let filter: any = { _id: new ObjectId(input.userId) }

    if (input.group) {
      filter = {
        ...filter,
        'roles.group': { $in: [ RoleService.GLOBAL, input.group ] }
      }
    }

    return this.mongoService.get<User>('users', filter, { fields: { roles: 1 } }).pipe(
      map(user => {
        if (user && user.roles) {
          // filter for global group
          // if group specified, filter by group
          const roles = user.roles
            .filter(role => role.group === RoleService.GLOBAL || !input.group || role.group === input.group)
            .map(role => role.role)

          return Array.from(new Set(roles))
        } else {
          return []
        }
      })
    )
  }

  getUsersInRoles (input: { roles: string[]; group?: string; limit?: number; cursor?: string }): Observable<FindOutput<User>> {
    let filter
    if (input.group) {
      filter = {
        roles: {
          $elemMatch: {
            role: { $in: input.roles },
            group: input.group
          }
        }
      }
    } else {
      filter = {
        'roles.role': { $in: input.roles }
      }
    }

    return this.mongoService.findWithCursor<User>('users', filter,
      input.limit, input.cursor, [ { field: 'profile.displayName', asc: true } ])
  }

  isUserInRoles (input: { userId: string; roles: string[]; group?: string }): Observable<boolean> {
    const filter = {
      _id: new ObjectId(input.userId),
      roles: {
        $elemMatch: {
          role: { $in: input.roles },
          group: {
            $in: input.group ? [ input.group, RoleService.GLOBAL ] : [ RoleService.GLOBAL ]
          }
        }
      }
    }

    return this.mongoService.findWithCursor<User>('users', filter, 1).pipe(
      map(users => users.docs.length > 0),
      catchError(e => {
        // whenever error happens, return as user not in role
        console.error(`Error checking if user [${input.userId}] is in roles`, e)
        return of(false)
      })
    )
  }

  removeRole (name: string): Observable<boolean> {
    return this.mongoService.get(
      'users',
      { 'roles.role': name },
      { fields: { _id: 1 } }
    ).pipe(
      concatMap(userWithRole => userWithRole ? of(false) :
        this.mongoService.remove('roles', { name }).pipe(
          map(remove => !!remove)
        ))
    )
  }

  removeUsersFromAllRoles (input: { userIds: string[] }): Observable<null> {
    const filter = {
      _id: { $in: input.userIds.map(id => new ObjectId(id)) }
    }
    const update = {
      $unset: {
        roles: ''
      }
    }

    return this.mongoService.collection('users').pipe(
      concatMap(users => users.updateMany(filter, update)),
      map(updateMany => {
        if (updateMany.result.ok === 1) {
          return null
        } else {
          throw new Error(`Error removing roles for users ${input.userIds}`)
        }
      })
    )
  }

  removeUsersFromRoles (input: { userIds: string[]; roles: string[]; group?: string }): Observable<null> {
    if (input.group) {
      return throwError('Unsupported input parameter [group] for removeUsersFromRoles()')
    }

    const filter = {
      _id: { $in: input.userIds.map(id => new ObjectId(id)) }
    }
    const update = {
      $pull: {
        roles: {
          role: { $in: input.roles }
        }
      }
    }

    return this.mongoService.collection('users').pipe(
      concatMap(users => users.updateMany(filter, update)),
      map(updateMany => {
        if (updateMany.result.ok === 1) {
          return null
        } else {
          throw new Error(`Error removing roles ${input.roles} for users ${input.userIds}`)
        }
      })
    )
  }

  role (id: string): Observable<Role | null> {
    return this.mongoService.get('roles', id)
  }

  updateRole (id: string, role: Partial<Role>): Observable<null> {
    return this.mongoService.update('roles', id, {
      $set: {
        modifiedDate: new Date(),
        name: role.name
      }
    })
  }
}
