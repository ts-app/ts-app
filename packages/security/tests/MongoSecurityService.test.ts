import { MongoService } from '@ts-app/mongo'
import { of } from 'rxjs'
import { catchError, concatMap, tap } from 'rxjs/operators'
import { MongoSecurityService, SecurityService } from '../src'
import { omitVolatile } from '../../common/src'

describe('MongoSecurityService', async () => {
  const localUrl = 'mongodb://localhost:27017'
  let mongoService: MongoService
  let securityService: SecurityService

  beforeEach(done => {
    mongoService = new MongoService(localUrl)
    securityService = new MongoSecurityService(mongoService)
    of('users').pipe(
      concatMap(collection => mongoService.dropCollection(collection)),
      catchError(e => {
        // just ignore if cannot drop
        console.debug(e)
        return of(null)
      })
    ).subscribe(
      undefined,
      e => console.error(e),
      () => done()
    )
  })

  afterEach(done => mongoService.close().subscribe(() => done()))

  test('user sign up & login with email/password (via bcrypt)', done => {
    const emailPassword = { email: 'test@test.com', password: 'abc123' }
    const badLogin = { email: 'test@test.com', password: 'abc123!' }

    // --- user sign up
    securityService.signUp(emailPassword).pipe(
      tap(signUp => {
        expect(signUp.error).toBeFalsy()
        const bcrypt = signUp.user!.services.password.bcrypt
        expect(bcrypt.length > 0).toBeTruthy()
        delete signUp.user!.services.password.bcrypt
        expect(omitVolatile(signUp.user!)).toMatchSnapshot()
      }),

      // --- bad login
      concatMap(() => securityService.loginWithEmailPassword(badLogin)),
      tap(login => expect(login).toMatchSnapshot()),

      // --- good login
      concatMap(() => securityService.loginWithEmailPassword(emailPassword)),
      tap(login => {
        // TODO: assert tokens
        expect(login.error).toBeFalsy()
        delete login.user!.services.password.bcrypt
        expect(omitVolatile(login.user!)).toMatchSnapshot()
      }),

      concatMap(login => securityService.user(login.user!.id)),
      tap(user => {
        delete user!.services.password.bcrypt
        expect(omitVolatile(user!)).toMatchSnapshot()
      }),

      // --- sign up for user that already exist
      concatMap(() => securityService.signUp(emailPassword)),
      tap(signUp => expect(signUp).toMatchSnapshot())
    ).subscribe(() => {
      expect.assertions(8)
      done()
    })
  })

  test('user sign up with invalid email', done => {
    securityService.signUp({
      email: 'bob',
      password: '123'
    }).pipe(
      tap(signUp => expect(signUp).toMatchSnapshot())
    ).subscribe(() => {
      expect.assertions(1)
      done()
    })
  })
})
