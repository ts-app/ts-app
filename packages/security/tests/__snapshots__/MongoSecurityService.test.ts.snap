// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoSecurityService user sign up & login with email/password (via bcrypt) 1`] = `
Object {
  "emails": Array [
    Object {
      "email": "test@test.com",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "test",
    "email": "test@test.com",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoSecurityService user sign up & login with email/password (via bcrypt) 2`] = `
Object {
  "error": Object {
    "message": "Invalid login attempt",
  },
}
`;

exports[`MongoSecurityService user sign up & login with email/password (via bcrypt) 3`] = `
Object {
  "emails": Array [
    Object {
      "email": "test@test.com",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "test",
    "email": "test@test.com",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoSecurityService user sign up & login with email/password (via bcrypt) 4`] = `
Object {
  "emails": Array [
    Object {
      "email": "test@test.com",
      "verified": true,
    },
  ],
  "profile": Object {
    "displayName": "test",
    "email": "test@test.com",
  },
  "services": Object {
    "password": Object {},
  },
}
`;

exports[`MongoSecurityService user sign up & login with email/password (via bcrypt) 5`] = `
Object {
  "error": Object {
    "message": "User already exist",
  },
}
`;

exports[`MongoSecurityService user sign up with invalid email 1`] = `
Object {
  "error": Object {
    "message": "Invalid email",
  },
}
`;
